File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix)
    {
        List<Integer> res = new ArrayList<>();
        int startrow=0;
        int startcol=0;
        int n=matrix.length;
        int m=matrix[0].length;
        int endrow=n-1;
        int endcol=m-1;
        while(startrow<=endrow&&startcol<=endcol)
        {
            for(int j=startcol;j<=endcol;j++)
            {
                res.add(matrix[startrow][j]);
            }
            for(int i=startrow+1;i<=endrow;i++)
            {
                res.add(matrix[i][endcol]);
            }
            for(int j=endcol-1;j>=startcol;j--)
            {
                if(startrow==endrow)
                {
                    break;
                }
                res.add(matrix[endrow][j]);
            }
            for(int i=endrow-1;i>=startrow+1;i--)
            {
                if(startcol==endcol)
                {
                    break;
                }
                res.add(matrix[i][startcol]);
            }
             

            startrow++;
            startcol++;
            endrow--;
            endcol--;
        }
        return res;
        
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {

        int rowstart=0;
        int rowend=matrix.size()-1;
        int colstart=0;
        int colend=matrix[0].size()-1;
        vector<int>ans;

     while(rowstart<=rowend&&colstart<=colend)
     {
        
        for(int i=colstart;i<=colend;i++)
          ans.push_back(matrix[rowstart][i]);
          
          rowstart++;
        
        
        
        for(int j=rowstart;j<=rowend;j++)
        
            ans.push_back(matrix[j][colend]);
            colend--;
        
      
       if(rowstart<=rowend)
       {
        for(int k=colend;k>=colstart;k--)
         ans.push_back(matrix[rowend][k]);
            rowend--;
       }
        
        if(colstart<=colend)
        {
        for(int l=rowend;l>=rowstart;l--)
           ans.push_back(matrix[l][colstart]);
         colstart++;
        }
        
     }

     return ans;   
    }
};